### PR TITLE
 Support Pulsar Message as function input

### PR DESCRIPTION
**Pull request by jerrypeng**
### Motivation

Support Message object as function input so that users can get all of the attributes of the Message.

Implementing this for windowed function is a more complicated and will discuss/implement after this

If everyone is ok with this approach then I will add tests

Discussions about this was done in:

https://github.com/apache/pulsar/pull/4127

### CODE REVIEW

**Comment by jerrypeng** [comment]
@ConcurrencyPractitioner please also take a look

File: pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/MessageInputFunction.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * under the License.
+ */
+package org.apache.pulsar.functions.api.examples;
+
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.functions.api.Context;
+import org.apache.pulsar.functions.api.Function;
+
+
+public class MessageInputFunction implements Function<Message<String>, String> {
```
Comment (sijie) [inline pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/MessageInputFunction.java]: Can you add an integration test for this?

Comment (jerrypeng) [reply pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/MessageInputFunction.java]: @sijie as mentioned in the description of the PR, once everyone is on board with this approach I will add the tests for this

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionCommon.java [hunk 2]
```diff
@@ -108,17 +110,51 @@ public static int findAvailablePort() {
                 Type actualInputType = ((ParameterizedType) collectionType).getActualTypeArguments()[0];
                 typeArgs[0] = (Class<?>) actualInputType;
             }
+            if(typeArgs[0].equals(Message.class)) {
```
Comment (srkukarni) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionCommon.java]: whats the plan for adding this support to window function?

Comment (jerrypeng) [reply pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionCommon.java]: we will probably have to change the interface from

```
public interface WindowFunction<I, O> {
    O process(Collection<Record<I>> input, WindowContext context) throws Exception;
}
```
to 

```
public interface WindowFunction<I, O> {
    O process(Collection<I> input, WindowContext context) throws Exception;
}
```

And support "I" as Record for BC and support "I" as Message.

File: pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/MessageInputFunction.java (hunk 1 above)
Comment (jerrypeng) [reply pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/MessageInputFunction.java]: @sijie I have added the tests

**Comment by ConcurrencyPractitioner** [comment]
Sure, I have no problem with this. Thanks for picking this up. :)

**Comment by jerrypeng** [comment]
rerun cpp tests

**Comment by jerrypeng** [comment]
@srkukarni @sijie can you look at this again? Thanks

**Comment by jerrypeng** [comment]
Alternatively, instead of allowing user's to be able to have Message<T> as input to a function, a simpler approach of we would be just to add this API to Context:

```
 Message getInternalMessage();
```

and the Impl would be:

```
Message getInternalMessage() {
return ((PulsarRecord) this.record).getMessage);
}
```

There also isn't any good way (without breaking BC) for the WindowFunction API to support Message as a input since the WindowFunction API is 

```
public interface WindowFunction<I, O> {
    /**
     * Process the input.
     *
     * @return the output
     */
    O process(Collection<Record<I>> input, WindowContext context) throws Exception;
}
```

I don't think it makes much sense to do 

```
    O process(Collection<Record<Message<I>>> input, WindowContext context) throws Exception;

```

**Comment by jerrypeng** [comment]
The simplest and non-BC-breaking approach I can think of for Functions and WindowFunctions:

just to add these getter methods to Context and WindowContext:

```
Message getInternalMessage();

```

and 

```
List<Message> getWindowAsMessages()
```

**Comment by merlimat** [comment]
@jerrypeng What about the message type? I believe `Context` is not generic at this point. Would you get a `Message<?>` then?

**Comment by jerrypeng** [comment]
@merlimat that is correct which is a con for that approach.

**Comment by jerrypeng** [comment]
@sijie @srkukarni @merlimat what we can do is this:
https://github.com/apache/pulsar/pull/4341

simpler approach that solves all our problems

**Comment by srkukarni** [comment]
+1 for #4341

**Pull request closed without merging**